pkg/dockerregistry/server/digestcache.go

```diff
@@ -52,42 +75,134 @@ func (c digestToRepositoryCache) RepositoriesForDigest(dgst digest.Digest) []str
 	return repos.Copy()
 }
 
+func (c digestToRepositoryCache) RepositoryHasBlob(repo string, dgst digest.Digest) bool {
+	value, ok := c.Get(dgst.String())
+	if !ok {
+		return false
+	}
+	repos := value.(*repositoryBucket)
+	return repos.Has(repo)
+}
+
+// repositoryBucket contains a list of repositories with eviction timeouts.
 type repositoryBucket struct {
-	mu   sync.Mutex
-	list []string
+	mu    sync.Mutex
+	clock util.Clock
+	list  []bucketEntry
 }
 
 // Has returns true if the bucket contains this repository.
-func (i *repositoryBucket) Has(repo string) bool {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	for _, s := range i.list {
-		if s == repo {
-			return true
+func (b *repositoryBucket) Has(repo string) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.evictStale()
+
+	return b.getIndexOf(repo) >= 0
+}
+
+// Add one or more repositories to this bucket.
+func (b *repositoryBucket) Add(ttl time.Duration, repos ...string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.evictStale()
+	evictOn := b.clock.Now().Add(ttl)
+
+	for _, repo := range repos {
+		index := b.getIndexOf(repo)
+		arr := b.list
+
+		if index >= 0 {
+			// repository already exists, move it to the end with highest eviction time
+			entry := arr[index]
+			copy(arr[index:], arr[index+1:])
+			if entry.evictOn.Before(evictOn) {
+				entry.evictOn = evictOn
+			}
+			arr[len(arr)-1] = entry
+
+		} else {
+			// repo is a new entry
+			if len(arr) >= bucketSize {
+				copy(arr, arr[1:])
+				arr = arr[:bucketSize-1]
```

Comment:
if the array is longer than the bucketSize, then we're chopping off the most recently added items here. probably not a big deal, since if we add correctly, we never exceed the bucketSize